### ISSUE TITLE
Add Maven-style fetching of dependencies while doing a build

### ISSUE THREAD

**Issue by kaby76** [kaby76/OldAntlr4BuildTasks#14]
This is an issue that was raised in [StackOverflow](https://stackoverflow.com/questions/62576310/preferred-way-to-bundle-shell-scripts-with-a-github-action) by @NiccoMlt .

When I forked Antlr4BuildTasks and added in code to call the Java Antlr tool, the build model departed from how this tool used to work. Rather than not requiring any dependencies to be downloaded and installed by the developer, Antlr4BuildTasks now required the developer to set up some environmental variables, and download and install Java and the Antlr complete jar toolkit. In contrast, when doing a build with Maven or CMake with Antlr, it fetches all dependencies.

Antlr4BuildTasks should follow the more modern paradigm.

This part should be easy to do. And, I already provide options to override the new default behavior.

However, extending this idea further, the build could download the source for the runtime, and build that. Antlr C++ programs that use CMake do this. Fortunately, building the C# runtime is very quick.

**Comment by kaby76** [comment]
I have the part that downloads dependencies working now--surprising myself that it can be done. But, there are a number of issues:

- You must have NET Core installed. This is because the only way for me to generate Antlr4.Runtime.Standard for 4.8 is to perform a build using the dotnet executable. (See https://github.com/kaby76/Antlr4BuildTasks/commit/2b98148b42fed1fbeec7db1915cc86cf536a5d28#diff-b084fccec07065ae602cfafe4fbec277). MSBuild is really a pain in the ass.
- Everything is currently hardwired to 4.8. I'll be abstracting this out.
- As far as I can tell, the only download that contains a working "java.exe" executable is the OpenJDK (see https://github.com/kaby76/Antlr4BuildTasks/commit/2b98148b42fed1fbeec7db1915cc86cf536a5d28#diff-11d16e67ea2e9b942e92dfb405b81489). This file, unfortunately, is a rippin' *slow* download, taking a few minutes from my ISP. I'm hoping I can find a Java installation that I can unzip that is smaller, and therefore faster to download.

**Comment by NiccoMlt** [comment]
>     * As far as I can tell, the only download that contains a working "java.exe" executable is the OpenJDK (see [2b98148#diff-11d16e67ea2e9b942e92dfb405b81489](https://github.com/kaby76/Antlr4BuildTasks/commit/2b98148b42fed1fbeec7db1915cc86cf536a5d28#diff-11d16e67ea2e9b942e92dfb405b81489)). This file, unfortunately, is a rippin' _slow_ download, taking a few minutes from my ISP. I'm hoping I can find a Java installation that I can unzip that is smaller, and therefore faster to download.
> 
>     * I only have this working on Windows.

Wait, why also downloading Java? I mean, ANTLR is a dependency, a JVM instance that runs it is not, in my opinion.

I agree with you that a way to pull official ANTLR jar from Maven is a great addition, as I also stated on StackOverflow,
but I think that shouldn't be a responsibility of your build task to also install the execution environment.

It is really easy for a user to add a JVM environment to the build environment (being it his/her own computer or a CI like GitHub Actions or Travis CI), so I think you could simply check if JAVA_HOME (or JAVA_EXEC, if you like it more) is set and warn the user with an error if it is not.
In the error you can easily link [Oracle's Java distribution](https://www.oracle.com/java/technologies/javase-downloads.html) or one of the many OpenJDK distributions (the [README of the Jabba tool](https://github.com/shyiko/jabba/blob/master/README.md) lists all the options).

**Comment by kaby76** [comment]
I want this to be easy for basic C# programmers, so that's why I define three dependencies that C# programmers are going to forget: Java, the Antlr tool jar, and the Antlr runtime. This is what I'm thinking:

- Java:
   - If JAVA_EXEC is empty, I first check if it's in $PATH by assuming "java.exe" when I do a spawn.
   - If JAVA_EXE is empty, and I can't find Java on $PATH, then I'll download the JDK (195MB).
   - If JAVA_EXE is not empty, I'll just spawn a process with that. If that doesn't work, fail, I guess.

- The Antlr4 Tool jar:
   - If Antlr4ToolPath is empty, I'll just download it (2MB).
   - If Antlr4ToolPath isn't empty, I'll use that and complain if it doesn't work.

- The Antlr4 runtime:
   - That's currently on Nuget. But, the versions have to agree between the tool and the runtime, something people are going to mess up. Also, it sometimes gets difficult to debug without the debug symbols for the runtime--the Nuget copy is only a Release build.
   - If @(ProjectReference) contains the Antlr4.Runtime.Standard, then I'll assume the person knows what he's doing, and not download, build and link the runtime.
   - If @(ProjectReference) does not contain the Antlr runtime, then I'll download (5MB), build, and link the runtime.

**Comment by NiccoMlt** [comment]
I understand that, and I think that wanting to make life easier to newbie developers is great.
Despite this, I think that the choice to handle the Java execution environment like a normal build dependency is not a great idea: calling into question the Single Responsibility Principle, I think that is not responsibility of this build task to configure Java environment. 
If you really want to install Java locally in a C# project, I think it would be a better idea to have a separate build task that does exactly this: fetch a JDK.

**Comment by kaby76** [comment]
For practical matters, I'm going to add the code to download Java because the OpenJDK file is NOT at all like any other basic developer software program where it must be "installed" in the usual sense. It's more akin to the antlr-4.8-complete.jar file. It is just downloaded and decompressed. I don't want to require anyone to mess up their development environment just for this build tool. The JDK is completely self-contained, and I can pick it to be exactly the same version that Parr used when he compiled and generated the JAR.

However, I will follow the following rules to avoid the lengthy download: if you set JAVA_EXEC set, I'll use that. If Java is "installed" and on $PATH, I'll use that. Rather than require the user download and stuff the JDK file in some directory, and call it an "installation", or require yet another copy of Java to be actually installed in the tradition of an installer, along with forcing the user to mess around with $PATH, I'll just use this self-contained JDK.

Alternatively, I could download this and the other dependencies to a standard temp area, and after the first build, you would never see a delay ever again.

Every one of these dependencies is already in a separate "buld task", i.e., Microsoft.Build.Utilities.Task, in the tool. It's currently in one fat DLL, but I may break that up if I saw that it could be reused elsewhere.

Have no fret. This will work great.

--Ken

**Comment by NiccoMlt** [comment]
> Have no fret. This will work great.

I know that JDK is not an "invasive" environment that requires s system-wide installation; many tools (like Jabba, SDKMAN! or even IDEs like IntelliJ IDEA) already allow local userspace "installations" of different versions of JDK (which is in fact putting the content of a zip into a folder), so I agree with you that you will probably be able to make it work pretty well.

My point was that ensuring Java is available shouldn't be a responsibility of a DLL that simply leverages ANTLR to generate C# sources; it adds complexity to your code and increases likelihood of potential future issues.
Ultimately, you are the maintainer of the software, so it's up to you to weigh that risk.

**Comment by kaby76** [comment]
The build package has been updated. Rather than set JAVA_EXEC and Antlr4ToolPath in the environment, download and install Java and the Antlr Tool jar, the package now requires only PackageReference's to Antlr4BuildTools v8.0 and Antlr4.Runtime.Standard 4.8 or 4.7.2. It will look at the version for the runtime in the .csproj file and pick the Antlr Tool jar file that is a match. Therefore, you don't need to download the jar. This clears up any problems of a version skew between the runtime and tool. Further, I created a small JRE that works with Antlr4.Runtime.Standard 4.8 built with Java 11 and jlink. So, Java does not need to be installed. If you want to override anything, use the old environmental variables as before. It only downloads Antlr4.Runtime.Standard 4.7.2 if you link to that. I don't provide it with this package. I investigated many ways of using the Antlr4ToolPath or making an assumption that it would be v4.8, then deriving the Antlr4.Runtime.Standard version to use. However, it turns out one cannot add a PackageReference on-the-fly in the .targets file because the NuGet restore is very very primitive. It does not execute the .targets file to gather package dependencies, and does not perform recursion to download further dependencies. This tool has now been moved to a new repository with the same name, and this forked repository renamed and archived. The tool has been rewritten so much so that it shares little with the original Harwell's tool, other than sharing the same purpose of supporting build rules for Antlr parsers.
